doc/DHCPv6_Relay/DHCPv6_Relay_HLD.md

```diff
@@ -0,0 +1,155 @@
+  - [2.3 DHCP Monitor](#23-dhcp-monitor)
+- [3 CLI](#3-cli)
+- [4 Init flow](#4-init-flow)
+
+# List of Tables
+* [Table 1: Abbreviations](#definitionsabbreviation)
+
+# List of Figures
+* [DHCPv6 Diagram](#2-modules-design)
+* [DHCPv6 init flow](#4-init-flow)
+
+# Revision
+| Rev | Date     | Author          | Change Description                 |
+|:---:|:--------:|:---------------:|------------------------------------|
+| 0.1 | 03/04    | Shlomi Bitton   | Initial version                    |
+
+# About this Manual
+This document provides an overview of the implementation and integration of DHCP Relay for IPv6 feature in SONiC.
+
+# Scope
+This document describes the high level design of the DHCP Relay for IPv6 feature in SONiC.
+
+# Definitions/Abbreviation
+| Abbreviation  | Description                               |
+|---------------|-------------------------------------------|
+| DHCP          | Dynamic Host Configuration Protocol       |
+
+# 1 Requirements Overview
+
+## 1.1 Functional Requirements
+
+DHCP Relay for IPv6 feature in SONiC should meet the following high-level functional requirements:
+
+- Give the support for relaying DHCP packets from downstream networks to upstream networks using IPv6 addresses.
+- Provide the functionality as a seperate process running on dhcp-relay docker container.
+- Relaying messages to multiple unicast and multicast addresses.
+
+## 1.2 Configuration and Management Requirements
+
+- DHCPv6 trap should be enabled through the COPP manager when the DHCP relay feature is enabled and vice versa. 
+- Downstream network is the VLAN interface with the relay configuration. Global IPv6 address is required to be configured on that interface. 
+- Config DB schema should meet the following format:
+```
+{
+"VLAN": {
+  "Vlan1000": {
+    "dhcp_servers": [
+      "192.0.0.1", 
+      "192.0.0.2", 
+    ], 
+    "dhcpv6_servers": [ 
+      "21da:d3:0:2f3b::7", 
+      "21da:d3:0:2f3b::6", 
+    ], 
+    "vlanid": "1000" 
+    } 
+  }
+}
+```
+
+# 2 Modules design
+
+![DHCPv6 Diagram](/doc/DHCPv6_Relay/diagram.png)
+
+## 2.1 DHCP Relay for IPv6 build and runtime dependencies
+
+The DHCP Relay for IPv6 feature, same as the IPv4 version, will be based on the open source project 'isc-dhcp'.
+
+## 2.2 DHCP Relay for IPv6 process in dhcp-relay container
+
+A new process will run in parallel to the other process for IPv4 support.
+The new process will listen to DHCP packets for IPv6 and forward them to the relevant interface according to the configuration.
+For example, from the configuration described on the previous section, the following daemon will start:
+```
+admin@sonic:/# /usr/sbin/dhcrelay -6 -d --name-alias-map-file /tmp/port-name-alias-map.txt -l Vlan1000 -u 21da:d3:0:2f3b::7%Ethernet28 -u 21da:d3:0:2f3b::6%Ethernet28
+```
+
+## 2.3 DHCP Monitor
+
+The existing DHCP monitor will be enhanced in order to support monitoring for DHCP IPv6 as well.
```

Comment:
DHCP counters will be part of the new DHCPv6 relay.

Reply:
So with the new approach the DHCP monitor application will be removed?

Reply:
Is there any support for DHCPv6 relay counters ? These will help a lot for debugging.

Reply:
Yes there is, the DHCP monitor application is now supporting monitor for these counters as well.
Please check the PR of the implementation.

Reply:
> Is there any support for DHCPv6 relay counters ? These will help a lot for debugging.

Yes, there will be native support in the app to pull counters. Please have a look at [PR:787](https://github.com/Azure/SONiC/pull/787)

Reply:
> Yes there is, the DHCP monitor application is now supporting monitor for these counters as well.
> Please check the PR of the implementation.

the dhcpmon will not exist in DHCPv6 world, rather its functionality will be in the new SONiC DHCPv6 RA.

Reply:
Can you please elaborate why the dhcpmon will not exist?
Same as the IPv4 implementation of this monitor, we can check inconsistency in received and transmitted packets.
for received SOLICIT/REQUEST packets we expect to see a RELAY-FORWARD packet transmitted.
for received RELAY-REPLY packets we expect to see ADVERTISE/REPLY packet transmitted.

Reply:
@shlomibitton The driving force to have a monitoring application (dhcpmon) was that ISC DHCP does not integrate well with SONIiC infra and its counters were not readily available via telemetry.. With the new DHCPv6 RA, it will integrate with the SONiC redis db and its counters will be available via telemetry and cli. So, we thought about eliminating it. 

One might argue that the monitoring app would still capture anomalies in DHCPv6 behavior as it is external to it. I am not sure if this is a design pattern we would like to follow as if we end with every app having a shadow app that monitors it, one can image the number of app would grow...

Also, dhcpmon is launching per interface which is inefficient to have a process for every interface that has dhcp relay enabled.

Please comment on [PR:787](https://github.com/Azure/SONiC/pull/787)

Reply:
@tahmed-dev thanks for the explanation.
On this case, what do you suggest to use as a monitor application with the isc-dhcp-relay package on this case?
This HLD is based on this implementation and we would like to have some kind of a monitor right?
On the other HLD on 'counters' section all I can see is the approach to keep the number of each message type but that's about it.

Reply:
@shlomibitton counters will be available via telemetry. The cloud would alert on counters if incrementing. Also, we can alternatively add checks similar to dhcpmon monitoring check and print to syslog. I believe Telemetry is maturing now and so the former solution should be adopted.

For this HLD, it is fine to have dhcpmon changes for the time being. This HLD and the accompanying code will be reverted once the new solution is merged.